errors.go

```diff
@@ -0,0 +1,25 @@
+package artifact
+
+import "errors"
+
+// ErrHTTPS is returned when a non-https url is involved in a redirect
+var ErrHTTPS = errors.New("only resources served over https are allowed")
+
+// ErrCorrupt is returned when an artifact is corrupt
+var ErrCorrupt = errors.New("corrupt resource")
+
+// ErrExpectedRedirect is returned when a redirect is expected but not received
+var ErrExpectedRedirect = errors.New("expected redirect")
+
+// ErrUnexpectedRedirect is returned when we expect a redirect but do not
+// receive one
+var ErrUnexpectedRedirect = errors.New("unexpected redirect")
+
+// ErrBadRedirect is returned when a malformed redirect is received.  Example
+// would be an empty Location: header or a Location: header with an invalid URL
+// as its value
+var ErrBadRedirect = errors.New("malformed redirect")
+
+// ErrBadOutputWriter is returned when the output writer passed into a function
+// was able to be checked for its size and it contained more than 0 bytes
+var ErrBadOutputWriter = errors.New("output writer is not empty")
```

Comment:
In general I feel there is a lot more value in an error that says "Resource xyz is corrupt" or "Unexpected redirect from url xyz" than the more general error "Corrupt resource" or "Unexpected redirect". When seen in logging, it enables you to troubleshoot much more effectively. For this reason I would prefer we have custom error types that include the instance data that caused the failure (the url that had the unexpected redirect, etc) and they can have a `func Error() string` method that returns a descriptive error containing that pertinent data. It is more work upfront, but saves a lot of time and work later. The type of error can still be detected in a [type switch](https://tour.golang.org/methods/16) statement.

Reply:
`error.go` has a lot of work now which produces errors like this:

```
Artifact Error:
  1. (internal) downloading http://127.0.0.1:40011/the-redirect to write to test-file-output728894125
  2. (internal) corrupt resource
```